libraries/apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/ir/IrOperationsBuilder.kt

```diff
@@ -751,20 +745,80 @@ internal class IrOperationsBuilder(
         visitedTypes.add(type)
 
         if (regexes.any { it.matches(type) }) {
-          usedFields.putType(type)
+          usedCoordinates.putType(type)
+
+          // Add all arguments of all fields
+          val gqlTypeDefinition = schema.typeDefinition(type)
+          gqlTypeDefinition.fieldDefinitions(schema).forEach { gqlFieldDefinition ->
+            gqlFieldDefinition.arguments.forEach { gqlArgumentDefinition ->
+              usedCoordinates.putArgument(gqlTypeDefinition.name, gqlFieldDefinition.name, gqlArgumentDefinition.name)
+              // Recursively add type of this argument
+              typesToVisit.add(gqlArgumentDefinition.type.rawType().name)
+            }
+          }
         }
 
         val gqlTypeDefinition = schema.typeDefinition(type)
         gqlTypeDefinition.fieldDefinitions(schema).forEach { gqlFieldDefinition ->
           if (regexes.any { it.matches("${gqlTypeDefinition.name}.${gqlFieldDefinition.name}") }) {
-            usedFields.putField(gqlTypeDefinition.name, gqlFieldDefinition.name)
+            usedCoordinates.putField(gqlTypeDefinition.name, gqlFieldDefinition.name)
             // Recursively add types used by this field
             typesToVisit.add(gqlFieldDefinition.type.rawType().name)
           }
         }
       }
     }
   }
+
+  private fun UsedCoordinates.putAllArguments(
+      parentType: String,
+      selections: List<GQLSelection>,
+      allFragmentDefinitions: Map<String, GQLFragmentDefinition>,
+  ) {
```

Comment:
If you end up keeping this, `allFragmentDefinitions` is in the context already, I think you can skip it here.

Also side note: yet another thing that would be easier to do with a proper traversal API.